docs: add documentation for kargo-specific annotations and labels (#3998)
akuity/kargo: docs/docs/50-user-guide/70-annotations.md

---
sidebar_label: Annotations
description: List of Kargo-specific annotations and labels
---

# Annotations and Labels used by Kargo

> This page documents Kargo-specific annotations, labels, and finalizers. Some of these fields can be set by users, while others are system-managed and used internally by the control plane.

Reviewer: I would reword this to say this documents a subset of annotations and labels that may be useful to advanced users. I would probably even go so far as to say that any annotations and labels used only internally and _not_ of any particular use to users are omitted.

Reviewer: Also, this whole page should move under the users/reference section of the doc tree. Keeping nitty gritty details like these under a reference section is a clear signal to users that it's documentation they can refer to on an as-needed basis and ought not be included in a quick read through of the docs.

## Annotations

> | Annotation key | Target resource(s) | Possible values | Description |
> |----------------|--------------------|------------------|-------------|

Reviewer: ```suggestion
| Key | Target Resource(s) | Possible Values | Description |
|-----|--------------------|-----------------|-------------|
```

> | `kargo.akuity.io/refresh` | `Warehouse`, `Stage`, `Promotion` | Arbitrary string (acts as a token) | Triggers reconciliation of the resource when the value changes. |

Reviewer: "Arbitrary string (acts as a token)" probably does not impart adequate guidance on how to use this. Maybe try something like this:

```suggestion
| `kargo.akuity.io/refresh` | `Warehouse`, `Stage`, `Promotion` | A string that is unique or at least unlikely to repeat, such as a UUID or a timestamp of "now" | Triggers reconciliation of the resource when its value changes. |
```

| `kargo.akuity.io/reverify` | `Stage` | Verification ID or JSON object (VerificationRequest) | Triggers re-verification of the current `Freight`. |

> | `kargo.akuity.io/abort` | `Stage` | Verification ID | Aborts an in-progress `Freight` verification. |

Reviewer: For verification ID, we probably need to tell users where they can find that. "JSON object (VerificationRequest)" doesn't really tell users much of anything. Even I don't know what the structure of that JSON should look like without first digging through the code.

Since these two rows probably require snippets of `StageStatus` represented as YAML in order to be properly informative, it makes me question whether a table is the best format for this information. You can try to make it look nice, but you may need to resort to making "Annotations" and "Labels" level 2 headings (i.e. `##`) and annotation/label keys level 3 (i.e. `###`).

> | `kargo.akuity.io/description` | Any | Free-form string | Human-readable description displayed in the UI. |

Reviewer: ```suggestion
| `kargo.akuity.io/description` | Any | Any string | Human-readable description displayed in the UI. |
```

> | `kargo.akuity.io/color` | `Stage` | Hex color code (e.g. `#ff8800`) | Optional cosmetic color used in UI. |

Reviewer: ```suggestion
| `kargo.akuity.io/color` | `Stage` | Hex color code (e.g. `#ff8800`) | Optional cosmetic color used in the UI's pipeline view. |
```

| `kargo.akuity.io/authorized-stage` | `Argo CD Application` | `<project>:<stage>` | Indicates which `Stage` is authorized to manage the `Application` resource. |

:::warning
Avoid setting system-managed annotations manually unless you're explicitly debugging internal behavior.

> :::

Reviewer: With the right changes [here](https://github.com/akuity/kargo/pull/3998/files#r2080295111), I think we can lose this whole warning box.

> ### RBAC Annotations

Reviewer: Since there are only two entries in the table below, I feel like "RBAC Annotations" is perhaps not worthy of its own distinct section. Let's just move these up into the previous section's table.

> | Annotation key | Target resource(s) | Possible values | Description |
> |----------------|--------------------|------------------|-------------|

Reviewer: ```suggestion
| Key | Target Resource(s) | Possible Values | Description |
|-----|--------------------|-----------------|-------------|
```

> | `rbac.kargo.akuity.io/managed` | `ServiceAccount`, `Role`, `RoleBinding` | `"true"` | Indicates that the resource is programmatically managed by Kargo's API server. Resources with this annotation are created and maintained by Kargo and should not be modified manually. |

Reviewer: ```suggestion
| `rbac.kargo.akuity.io/managed` | `ServiceAccount`, `Role`, `RoleBinding` | `"true"` |  Permits the UI or CLI (via the API server) to programmatically manage trios of `ServiceAccount`, `Role`, and `RoleBinding` resources via Kargo's own ["roles" abstraction](../50-user-guide/50-security/20-access-controls/index.md#managing-mappings-and-permissions). Omit this annotation if you wish to exclusively manage these resources [declaratively](../50-user-guide/50-security/20-access-controls/index.md#managing-kargo-roles-declaratively). |
```

> | `rbac.kargo.akuity.io/claim.<name>` | `ServiceAccount` | Claim name as suffix | Maps an OIDC claim (e.g., `sub`, `email`, or `groups`) to a `ServiceAccount`, enabling user-to-ServiceAccount mappings. For more details, see the Access Control section here: [Operator Guide](../40-operator-guide/40-security/30-access-controls.md), [User Guide](../50-user-guide/50-security/20-access-controls/index.md). |

Reviewer: ```suggestion
| `rbac.kargo.akuity.io/claim.<name>` | `ServiceAccount` | Claim name as suffix | Maps an OIDC claim (e.g., `sub`, `email`, or `groups`) to a `ServiceAccount`, enabling user-to-ServiceAccount mappings. For more details, refer to the access control sections of the [Operator Guide](../40-operator-guide/40-security/30-access-controls.md) and [User Guide](../50-user-guide/50-security/20-access-controls/index.md). |
```

## Labels

> | Label key | Target resource(s) | Possible values | Description |
> |-----------|--------------------|------------------|-------------|

Reviewer: ```suggestion
| Key | Target Resource(s) | Possible Values | Description |
|-----|--------------------|-----------------|-------------|
```

> | Label key | Target resource(s) | Possible values | Description |
> |-----------|--------------------|------------------|-------------|

Reviewer: ```suggestion
| Key | Target Resource(s) | Possible Values | Description |
|-----|--------------------|-----------------|-------------|
```

> | `kargo.akuity.io/alias` | `Freight` | User-defined string | Friendly identifier for the `Freight` version. This label is automatically synced from the `alias` field of the `Freight` resource. Users should modify the `alias` field directly rather than editing this label, as the label exists primarily for convenience when using tools like `kubectl`. |

Reviewer: ```suggestion
| `kargo.akuity.io/alias` | `Freight` | Any string that is unique within the project | Mutable, human-readable alias for a piece of `Freight`. This label is automatically synced from the resource's `alias` field. Users are discouraged from modifying the label directly.  The label exists primarily to enable querying for `Freight` by alias using `kubectl`. |
```

> | `kargo.akuity.io/project` | Project namespaces | `"true"` | Indicates that the namespace is eligible for adoption by a `Project` with the same name. This label is required for pre-created namespaces to be associated with a `Project`. For more details, see the [Working with Projects](./20-how-to-guides/20-working-with-projects.md#namespace-adoption) section. |

Reviewer: ```suggestion
| `kargo.akuity.io/project` | `Namespace` | `"true"` | Indicates that the `Namespace` is eligible for adoption by a `Project` with the same name. This label is useful when `Namespace`s are unavoidably pre-created by some other agent. For more details, see the [Working with Projects](./20-how-to-guides/20-working-with-projects.md#namespace-adoption) section. |
```

> | `kargo.akuity.io/freight-collection` | `Freight` | Collection name | System-assigned label that establishes a relationship between an `AnalysisRun` and a collection of `Freight`. |

Reviewer: This one feels relatively unimportant. Let's drop it.

```suggestion
```

> | `kargo.akuity.io/cred-type` | Credentials | `git`, `helm`, `image`, `generic` | Indicates the type of credentials stored in a `Secret` resource. This label is used to classify credentials for Git, Helm chart, container image repositories, or other generic purposes. For more details, see the [Managing Credentials](./50-security/30-managing-credentials.md#repository-credentials-as-secret-resources) section. |

Reviewer: ```suggestion
| `kargo.akuity.io/cred-type` | `Secret` | `git`, `helm`, `image`, `generic` | Indicates a `Secret` represents credentials for a repository of the specified type. For more details, see the [Managing Credentials](./50-security/30-managing-credentials.md#repository-credentials-as-secret-resources). |
```

> | `kargo.akuity.io/shard` | Internal resources | Shard ID | System-assigned label that indicates the logical cluster or controller shard ID. For `Warehouse` and `Stage` resources, users should modify the `spec.shard` field instead of this label. For `Promotions`, the system automatically copies the `Stage`'s `spec.shard` value to this label. |

Reviewer: idk that I would classify this as system-assigned. _Users_ decide that a Warehouse or Stage belong to a particular shard.

## Finalizers

| Key | Target resource(s) | Possible values | Description |
|-----|--------------------|------------------|-------------|

> | `kargo.akuity.io/finalizer` | All | N/A | Used by control plane to enforce proper cleanup. |

Reviewer: I am questioning whether this is useful information. It's certainly not something users should ordinarily be touching.

> :::info
> - If you're setting annotations via `kubectl`, use the `--overwrite` flag when updating existing annotations: 
>   ```
>   kubectl annotate stage my-stage kargo.akuity.io/refresh=$(date +%s) --overwrite
>   ```
> :::

Reviewer: This is good advice, but feels like an awkward thing to tack on to the end of this page given that nowhere else on the page did we include any instructions for adding/modifying annotations via `kubectl`.

Let's cut this and we can always put it back in the future if we discover this is something we really _should_ have said after all.

```suggestion
```

---
sidebar_label: Annotations
description: List of Kargo-specific annotations and labels
---

# Annotations and Labels used by Kargo

This page documents Kargo-specific annotations, labels, and finalizers. Some of these fields can be set by users, while others are system-managed and used internally by the control plane.

## Annotations

| Annotation key | Target resource(s) | Possible values | Description |
|----------------|--------------------|------------------|-------------|

> | `kargo.akuity.io/refresh` | Any Kargo resource | Arbitrary string (acts as a token) | Triggers reconciliation when the value changes. Useful for forcing a resync. |

Reviewer: "Any Kargo resource" isn't really true. Off the top of my head, I think this works only on Warehouses, Stages, and Promotions.

This sounds like it's just saying the same thing twice:

> Triggers reconciliation when the value changes. Useful for forcing a resync.

> | `kargo.akuity.io/reverify` | `Stage` | Verification ID or JSON object (VerificationRequest) | Triggers re-verification of the currently assigned Freight. |

Reviewer: "assigned Freight" sounds awkward. Freight gets _promoted_ to a Stage. We don't really speak in terms of Freight being _assigned_ to a Stage. I think the easiest resolution here would be this:

```suggestion
| `kargo.akuity.io/reverify` | `Stage` | Verification ID or JSON object (VerificationRequest) | Triggers re-verification of the current Freight. |
```

| `kargo.akuity.io/abort` | `Stage` | Verification ID | Aborts an in-progress Freight verification. |

> | `kargo.akuity.io/description` | Any | Free-form string | Human-readable description displayed in the UI. |

Reviewer: I'm not sure that the UI actually honors this for _any_ resource type as has been indicated here.

| `kargo.akuity.io/color` | `Stage` | Hex color code (e.g. `#ff8800`) | Optional cosmetic color used in UI. |

> | `kargo.akuity.io/create-actor` | All | Actor identity (username/service) | Injected by the control plane to indicate who created the resource. |

Reviewer: I'm questioning, especially given the admonishment later on this same page _not_ to mess with system-assigned annotations, whether this isn't actually information that we're better off omitting, at least until such time we decide there's greater value in documenting them.

For now, I'd cut this and we can revisit later if need be.

> | `kargo.akuity.io/authorized-stage` | `Freight`, `Promotion`, etc. | `<project>:<stage>` | Indicates which Stage is authorized to manage the resource. |

Reviewer: This annotation is actually only applied to Argo CD Application resources and nothing else.

Reviewer: https://deploy-preview-3998.docs.kargo.io/user-guide/how-to-guides/argo-cd-integration#authorizing-updates

> | `kargo.akuity.io/promotion` | `PromotionTask`, internal resources | Promotion ID or name | Links the resources to a promotion process. |

Reviewer: This would neither appear on a `PromotionTask` nor on _most_ resource types.

It doesn't link a resource to a "promotion process." It links a resource to a _Promotion_.

This is also another one that's system-assigned and may be better to omit until there's a solid reason not to.

> | `kargo.akuity.io/argocd-context` | `Stage` | JSON or internal identifier | Records Argo CD application context from the last promotion. |

Reviewer: Another one that is system-assigned.

:::warning
Avoid setting system-managed annotations manually unless you're explicitly debugging internal behavior.
:::

### RBAC Annotations

| Annotation key | Target resource(s) | Possible values | Description |
|----------------|--------------------|------------------|-------------|

> | `rbac.kargo.akuity.io/managed` | RBAC resources | `"true"` | Marks a resource as managed by Kargo's RBAC system. |

Reviewer: This is added to a _ServiceAccount_, Role, or RoleBinding. ServiceAccounts aren't RBAC resources. I know that may not seem to be the case, but ServiceAccount is part of the `v1` API and not the `rbac.authorization.k8s.io/v1` API.

This is also system-assigned when a "Kargo role" (a trio of SA, Role, RoleBinding resources) is created programmatically via the Kargo API server. The API server refuses to modify SAs, Roles, or RoleBindings that _aren't_ annotated this way. It's _permission_ for the API server to manage them. These annotations would/should be absent in the event of someone managing SAs, Roles, or RoleBindings declaratively, e.g. via GitOps, because they'd _not_ ever want them modified programmatically by Kargo.

So this is another case where I question how much a user needs to know this. I could see some value in describing what this does a bit more than I can see in the case of other system-assigned annotations. If we think it's worth keeping at all, then I think _more_ detail is necessary.

> | `rbac.kargo.akuity.io/claim.<name>` | RBAC resources | Claim name as suffix | Maps an OIDC claim to a policy. |

Reviewer: This annotation is only used on ServiceAccounts, which are not RBAC resources.

I would also avoid the term "policy," since that term doesn't accurately describe a ServiceAccount.

This annotation is already well-documented [here](https://deploy-preview-3998.docs.kargo.io/operator-guide/security/access-controls#user-to-serviceaccount-mappings) and [here](https://docs.kargo.io/user-guide/security/access-controls/#user-to-serviceaccount-mappings). Please read up to understand how this works and try to improve the accuracy of this row. You can also link from here to those more thorough explanations of this annotation's effects.

### Event Annotations

The following annotations are added internally by Kargo's eventing subsystem for tracking and audit. They are **not user-settable**.

| Annotation key | Target resource(s) | Possible values | Description |
|----------------|--------------------|------------------|-------------|
| `event.kargo.akuity.io/actor` | Events | Actor string | Actor that triggered the event. |
| `event.kargo.akuity.io/project` | Events | Project name | Associated project. |
| `event.kargo.akuity.io/promotion-name` | Events | Promotion name | Related promotion. |
| `event.kargo.akuity.io/promotion-create-time` | Events | Timestamp | Creation time of the promotion. |
| `event.kargo.akuity.io/freight-*` | Events | Varies | Additional metadata about Freight. |
| `event.kargo.akuity.io/stage-name` | Events | Stage name | Target Stage. |
| `event.kargo.akuity.io/analysis-run-name` | Events | Name | Linked Argo CD `AnalysisRun`. |
| `event.kargo.akuity.io/verification-*` | Events | Timestamps/status | Tracks verification lifecycle. |

> | `event.kargo.akuity.io/applications` | Events | Comma-separated names | Applications involved in promotion. |

Reviewer: I would leave these out until we think there's a reason not to.

## Labels

| Label key | Target resource(s) | Possible values | Description |
|-----------|--------------------|------------------|-------------|

> | `kargo.akuity.io/alias` | `Freight` | User-defined string | Friendly identifier for the Freight version. |

Reviewer: Freight has an `alias` field and its value gets synced to this label. (Only because the existence of the label is beneficial to `kubectl` users.) It's preferable for people to work with that field and not touch the label directly. This probably needs to be mentioned.

> | `kargo.akuity.io/stage` | `Freight`, `Promotion`, etc. | Stage name | Indicates the associated Stage. |

Reviewer: Another system-assigned one.

> | `kargo.akuity.io/project` | All | Project name | Denotes Project ownership. |

Reviewer: That is not what this label does. This label should be applied to Project namespaces with a value of `true` if they are created in advance (not by Kargo), so that the namespace can be "adopted" by a Project having the same name. This is covered in more detail [here](https://docs.kargo.io/user-guide/how-to-guides/working-with-projects#namespace-adoption). Please read up on it to improve the accuracy of this row and you can also link to it from here.

> | `kargo.akuity.io/freight-collection` | `Freight` | Collection name | Source collection from which Freight originated. |

Reviewer: That is not what this label is for. This is system-assigned to establish a relationship between an AnalysisRun and a collection of Freight.

> | `kargo.akuity.io/cred-type` | Credentials | Credential type (e.g., `ssh`, `https`) | Internal marker for credential classification. |

Reviewer: Possible values are actually `git`, `helm`, `image`, and `generic`.

https://docs.kargo.io/user-guide/security/managing-credentials#repository-credentials-as-secret-resources

> | `kargo.akuity.io/shard` | Internal resources | Shard ID | Logical cluster or controller shard ID. |

Reviewer: Similar to the alias label on Freight, users probably shouldn't touch this and should favor the `spec.shard` field on a Warehouse or Stage.

For Promotions, the system copies the Stage's `spec.shard` value to the `kargo.akuity.io/shard` label.
